test/dynamo/test_activation_checkpointing.py

```diff
@@ -1994,6 +1999,639 @@ def forward(self, primals_1: "f32[4, 4]"):
+    ones_like = torch.ops.aten.ones_like.default(sum_1, pin_memory = False, memory_format = torch.preserve_format);  sum_1 = None
+    expand = torch.ops.aten.expand.default(ones_like, [4, 4]);  ones_like = None
+    cos = torch.ops.aten.cos.default(arg0_1);  arg0_1 = None
+    mul = torch.ops.aten.mul.Tensor(expand, cos);  expand = cos = None
+    detach = torch.ops.aten.detach.default(mul);  mul = None
+    return (detach,)""",
+        )
+
+        # Verify correctness
+        self.assertTrue(torch.allclose(dx1, dx2))
+
+        # sin is used in forward (for sum), so it stays in forward
+        # But DCE-based approach still works correctly
+        order_with = self._get_node_order(gm_with)
+        first_bwd_idx = min(order_with[n] for n in self._get_backward_nodes(gm_with))
+        ac_in_fwd = sum(
+            1 for ac in self._get_ac_nodes(gm_with) if order_with[ac] < first_bwd_idx
+        )
+        # sin is needed for forward, so it's kept (DCE doesn't remove it)
+        self.assertEqual(ac_in_fwd, 1)
+
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
+    def test_ac_rematerialize_graph_structure(self):
+        """Verify graph structure with AC reordering enabled."""
+
+        x_data = torch.randn(4, 4)
+        y_data = torch.randn(4, 4)
+
+        def simple_fwd_bwd():
+            x = x_data.detach().requires_grad_(True)
+            y = y_data.detach().requires_grad_(True)
+            z = torch.utils.checkpoint.checkpoint(
+                lambda a, b: torch.matmul(a, b), x, y, use_reentrant=False
+            )
+            loss = z.sum()
+            with torch.fx.traceback.annotate({"backward": 0}):
+                dx, dy = _grad(loss, (x, y))
+            return dx.detach(), dy.detach()
+
+        _, captured_gm = self._compile_and_capture(simple_fwd_bwd, True)
+
+        # mm used in forward only (sum consumes it), DCE removes it
+        self.assertExpectedInline(
+            captured_gm.code.strip(),
+            """\
+def forward(self, arg0_1, arg1_1):
+    mm = torch.ops.aten.mm.default(arg0_1, arg1_1)
+    sum_1 = torch.ops.aten.sum.default(mm);  mm = None
+    ones_like = torch.ops.aten.ones_like.default(sum_1, pin_memory = False, memory_format = torch.preserve_format);  sum_1 = None
+    expand = torch.ops.aten.expand.default(ones_like, [4, 4]);  ones_like = None
+    t = torch.ops.aten.t.default(arg0_1);  arg0_1 = None
+    mm_1 = torch.ops.aten.mm.default(t, expand);  t = None
+    t_1 = torch.ops.aten.t.default(arg1_1);  arg1_1 = None
+    mm_2 = torch.ops.aten.mm.default(expand, t_1);  expand = t_1 = None
+    detach = torch.ops.aten.detach.default(mm_2);  mm_2 = None
+    detach_1 = torch.ops.aten.detach.default(mm_1);  mm_1 = None
+    return (detach, detach_1)""",
+        )
+
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
+    def test_ac_rematerialize_duplicates_nodes_used_in_both_regions(self):
```

Comment:
Should we keep just one of this test and test_ac_rematerialize_simple_forward_backward, or are they trying to test two different things.